core/dataflow/src/main/java/edu/uci/ics/texera/dataflow/udf/UserDFOperator.java

```diff
@@ -0,0 +1,281 @@
+    private boolean getPythonResult = false;
+    public Process processPython;
+    private String pythonPID;
+    
+    public Tuple outputTuple;
+    
+    private MappedByteBuffer inputBuffer;
+    private MappedByteBuffer outputBuffer;
+    
+    private FileChannel inputFileChannel;
+    private FileChannel outputFileChannel;
+    
+    public UserDFOperator(UserDFOperatorPredicate predicate) {
+        this.predicate = predicate;
+        this.pythonPID = null;
+        this.inputBuffer = null;
+        this.outputBuffer = null;
+    }
+    
+    private String getJavaPID() {
+        String vmName = ManagementFactory.getRuntimeMXBean().getName();
+        int p = vmName.indexOf("@");
+        return vmName.substring(0, p);
+    }
+    
+    @SuppressWarnings({ "restriction", "resource" })
+    private boolean handShake() {
+        // This function will initiate the communication between Java and the launched process.
+        try {
+            File inputFile = new File(inputFilePath);
+            //Delete the file; we will create a new file
+            inputFile.delete();
+            // Get file channel in readwrite mode
+            inputFileChannel = new RandomAccessFile(inputFile, "rw").getChannel();
+            
+            // Get direct byte buffer access using channel.map() operation,    320*K bytes
+            inputBuffer = inputFileChannel.map(FileChannel.MapMode.READ_WRITE, 0, predicate.mmapBufferSize);
+            
+            inputBuffer.position(0);
+            inputBuffer.put((getJavaPID()+"\n").getBytes());
+            
+            //build output Buffer mmap
+            File outputFile = new File(outputFilePath);
+            outputFile.delete();
+            outputFileChannel = new RandomAccessFile(outputFile, "rw").getChannel();
+            
+            outputBuffer = outputFileChannel.map(FileChannel.MapMode.READ_WRITE, 0, predicate.mmapBufferSize);
+            startPythonProcess();
+            while ( true ) {
+                Thread.sleep(200);
+                if (getPythonResult) {
+                    getPythonResult = false;
+                    break;
+                }
+            }
+            pythonPID = getPythonPID();
+            
+            if (pythonPID == null || pythonPID.length() == 0) {
+                return false;
+            }
+        } catch (Exception e) {
+            throw new TexeraException("Hands shaking Failed!");
+        }
+        return true;
+    }
+    
+    
+    /* Return one of the indications: TAG-WAIT, TAG_NULL or TAG_LEN
+     * */
+    public String getPythonPID() {
+        return readStringFromMMap(outputBuffer, predicate.POSITION_PID);
+    }
+    
+    public String getTagFromOutputBuffer() {
+        String strTag = "";
+        outputBuffer.position(predicate.POSITION_TAG);
+        while (true)
+        {
+            char ch;
+            if ((ch = (char) outputBuffer.get()) == Character.UNASSIGNED) {
```

Comment:
In Character.Unassigned case, the tag returned is an empty string?

Reply:
Also, what is the TAG for waiting, Character.Unassigned or "w"? It seems from the code that it is "w", then what is the use of Character.Unassigned?